termwiz/src/escape/apc.rs

```diff
@@ -246,14 +246,134 @@ impl KittyImageData {
 
                 Ok(data)
             }
-            Self::SharedMem { .. } => {
-                log::error!("kitty image protocol via shared memory is not supported");
-                Err(std::io::ErrorKind::Unsupported.into())
-            }
+            Self::SharedMem {
+                name,
+                data_offset,
+                data_size,
+            } => read_shared_memory_data(&name, data_offset, data_size),
         }
     }
 }
 
+#[cfg(unix)]
+fn read_shared_memory_data(
+    name: &str,
+    data_offset: Option<u32>,
+    data_size: Option<u32>,
+) -> std::result::Result<std::vec::Vec<u8>, std::io::Error> {
+    use nix::sys::mman::{shm_open, shm_unlink};
+    use std::fs::File;
+    use std::os::unix::io::FromRawFd;
+
+    let raw_fd = shm_open(
+        name,
+        nix::fcntl::OFlag::O_RDONLY,
+        nix::sys::stat::Mode::empty(),
+    );
+    if raw_fd.is_err() {
+        log::warn!("Unable to open shared memory object {}", name);
+        return Err(std::io::Error::last_os_error());
+    };
+    let mut f = unsafe { File::from_raw_fd(raw_fd.unwrap()) };
+    if let Some(offset) = data_offset {
+        f.seek(std::io::SeekFrom::Start(offset.into()))?;
+    }
+    let data = if let Some(len) = data_size {
+        let mut res = vec![0u8; len as usize];
+        f.read_exact(&mut res)?;
+        res
+    } else {
+        let mut res = vec![];
+        f.read_to_end(&mut res)?;
+        res
+    };
+
+    if let Err(err) = shm_unlink(name) {
+        log::error!(
+            "Unable to unlink kitty image protocol shm file {}: {:#}",
+            name,
+            err
+        );
+    }
+    Ok(data)
+}
+
+#[cfg(windows)]
+fn read_shared_memory_data(
+    name: &str,
+    data_offset: Option<u32>,
+    data_size: Option<u32>,
+) -> std::result::Result<std::vec::Vec<u8>, std::io::Error> {
+    use winapi::um::handleapi::CloseHandle;
+    use winapi::um::memoryapi::{
+        MapViewOfFile, OpenFileMappingW, UnmapViewOfFile, VirtualQuery, FILE_MAP_ALL_ACCESS,
+    };
+    use winapi::um::winnt::MEMORY_BASIC_INFORMATION;
+    /// Convert a rust string to a windows wide string
+    fn wide_string(s: &str) -> Vec<u16> {
+        use std::os::windows::ffi::OsStrExt;
+        std::ffi::OsStr::new(s)
+            .encode_wide()
+            .chain(std::iter::once(0))
+            .collect()
+    }
+    let wide_name = wide_string(&name);
+
+    let handle = unsafe { OpenFileMappingW(FILE_MAP_ALL_ACCESS, 0, wide_name.as_ptr()) };
+    if handle.is_null() {
+        log::warn!("Unable to open shared memory object {}", name);
+        return Err(std::io::Error::last_os_error());
+    }
+    let buf = unsafe { MapViewOfFile(handle, FILE_MAP_ALL_ACCESS, 0, 0, 0) };
+    if buf.is_null() {
+        unsafe {
+            CloseHandle(handle);
+        }
```

Comment:
It's a bit tricky to manually manage cleaning things up around file mappings.
Please adopt an approach more similar to this other code in wezterm:

https://github.com/wez/wezterm/blob/9a324385eb514fa2c04d74ced5325250bec3cd72/wezterm-client/src/discovery.rs#L39-L44

https://github.com/wez/wezterm/blob/9a324385eb514fa2c04d74ced5325250bec3cd72/wezterm-client/src/discovery.rs#L152-L156

The key benefit is that the mapping and view have Drop impl's that will ensure that they are released when the scope unwinds; that makes the logic much simpler and "guarantees" that we can't leak those resources.

Reply:
I've created a `SharedMemObject` with `Drop` impl that does `UnmapViewOfFile` and `CloseHandle`, but before the construction of this object, there is still one call to `CloseHandle` if mapping the handle fails. Let me know if this current implementation is acceptable, or if I need to create another struct wrapper for `HANDLE` and impl `Drop` for it to avoid calling `CloseHandle` outside of `Drop`.

Reply:
Thanks!

> I need to create another struct wrapper for `HANDLE` and impl `Drop` for it to avoid calling `CloseHandle` outside of `Drop`.

Please do: it's more robust in the face of refactoring and easier for humans if we don't have to remember the details around releasing resources!

Reply:
Have created a separate struct for `HANDLE` with `Drop` implemented for it, let me know if any further changes desired!